torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,432 @@
+        }
+        remaining_size -= example_count;
+      }
+    }
+
+    // If we still have data remaining after filling the last pushed batch, add
+    // them to the queue too.
+    while (remaining_size > 0) {
+      UnwrappedBatchType current_batch;
+
+      // Allocate the batch memory ahead of time.
+      current_batch.reserve(batch_size_);
+
+      auto example_count = std::min(remaining_size, batch_size_);
+      auto batch_example_indices = example_sampler_.next(example_count);
+      AT_ASSERT(
+          batch_example_indices &&
+          batch_example_indices.value().size() == example_count)
+      BatchRequestType indices = batch_example_indices.value();
+      for (size_t i : indices) {
+        current_batch.emplace_back(std::move(data[i]));
+      }
+      remaining_size -= example_count;
+      UnwrappedBatchData<UnwrappedBatchType> batch_data(
+          std::move(current_batch));
+      batch_queue_.push(std::move(batch_data));
+    }
+    this->total_example_count_in_queue_ += data_size;
+
+    AT_ASSERT(remaining_chunk_count_ > 0);
+    this->remaining_chunk_count_--;
+
+    lock.unlock();
+    cvr_.notify_all();
+  }
+
+  /// Push exceptions throwed during preloading into batch queue. Called from
+  /// the ChunkDataSet worker threads.
+  void add_chunk_data(size_t index, std::exception_ptr e_ptr) {
+    std::unique_lock<std::mutex> lock(queue_mutex_);
+    cvw_.wait(lock, [this] {
+      // stop loading if we have preloaded enough data.
+      return this->total_example_count_in_queue_ < queue_depth_;
+    });
+    UnwrappedBatchData<UnwrappedBatchType> batch_data(e_ptr);
+    batch_queue_.push(std::move(batch_data));
+
+    AT_ASSERT(remaining_chunk_count_ > 0);
+    remaining_chunk_count_--;
+    lock.unlock();
+    cvr_.notify_all(); // notify all readers.
+  }
+
+  /// count of remaining chunk to be loaded. It is initialized with the total
+  /// chunk count and it decreases when a chunk data is retrieved. When this reaches
+  /// to 0, no more chunk needs to be loaded.
+  size_t remaining_chunk_count_ = 0;
+
+  /// The batch size is needed to create batches from the chunk data. Similar to
+  /// regular dataloader where the batches are created with prefetches,
+  /// BatchDataBuffer perform the batch creation using the provided batch size.
+  size_t batch_size_ = 0;
+
+  /// count of total example stored in the queue
+  size_t total_example_count_in_queue_ = 0;
+
+  /// local cache to store example batches from loaded chunk
+  std::queue<UnwrappedBatchData<UnwrappedBatchType>> batch_queue_;
+
+  // sync batch_queue_ update.
+  std::mutex queue_mutex_;
+
+  std::condition_variable cvr_;
+  std::condition_variable cvw_;
+
+  ExampleSampler example_sampler_;
+
+  // indicator for whether an empty chunk should be ignored. When it is true, an
+  // example will throw, otherwise, this empty chunk is skipped.
+  bool ignore_empty_chunk_ = false;
+
+  // configurable queue depth for batch_queue_
+  size_t queue_depth_;
```

Comment:
nit: I don't think it's very common to call the number of elements in a queue its *depth*. It's not a hierarchical structure, so *size* seems more appropriate.

Reply:
another suggestion is to call it refill_threshold. It is not queue_size.

Reply:
Or queue_capacity_ :)